### DISCUSSION TITLE
Supabase storage with buffers or streams

### DISCUSSION THREAD

**Discussion by coderinblack08** [supabase/supabase#1496] [Questions] [answered]
I'm using sharp in a nextjs API route to compress and resize an image. Then I fetch the new image, in the form of a buffer. How would I go saving this to supabase storage (like an example I guess)? I've tried converting the Buffer to a blob and then to a file but it doesn't work. Should I just use the bare API rather than going through `supabase/storage-js`

**Comment by easy-street** [comment, marked as answer]
yes, looks like interacting with the API directly is the way to go for now, until the client supports this use case (per this comment:  https://github.com/supabase/storage-js/issues/5#issuecomment-831642601)

in a Next.js API route, this might look something like:
```
import FormData from "form-data";
import sharp from "sharp";

import { supabase } from "../../utils/initSupabase";

export default async function upload(req, res) {
  try {
    const fileName = "bernie.png";
    const formData = new FormData();

    formData.append(
      "",
      await sharp("/tmp/bernie.jpg")
        .resize(200)
        .png({ compressionLevel: 7 })
        .toBuffer(),
      fileName
    );
    formData.append("cacheControl", 3600);

    res.json(
      await fetch(`${supabase.storage.url}/object/images/${fileName}`, {
        method: "POST",
        headers: supabase.storage.headers,
        body: formData,
      }).then((response) => response.json())
    );
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
}
```

make sure to [Add Security Rules](https://supabase.io/docs/guides/storage#add-security-rules). the above works with this configuration:

![image](https://user-images.githubusercontent.com/64047794/117557908-a1dc6000-b03d-11eb-8a76-7f151f133f86.png)

or alternatively, create a new client with the `secret` API key:
```
import { createClient } from "@supabase/supabase-js";

const supabase = createClient(
  process.env.NEXT_PUBLIC_SUPABASE_URL,
  process.env.SUPABASE_SECRET_KEY
);
```

**Reply by coderinblack08** [reply]
Wow! Thanks for the response; I'll check it out for sure

**Reply by coderinblack08** [reply]
@easy-street The thing is, isn't `supabase.storage.headers` protected so I can't access it?

**Reply by easy-street** [reply]
yeah, typescript might yell about it. instead, maybe something like this for typescript:

```
import FormData from "form-data";
import sharp from "sharp";

import type { NextApiRequest, NextApiResponse } from "next";

const apikey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY;

export default async function upload(
  req: NextApiRequest,
  res: NextApiResponse
) {
  try {
    const fileName = "bernie.png";
    const formData = new FormData();

    formData.append(
      "",
      await sharp("/tmp/bernie.jpg")
        .resize(200)
        .png({ compressionLevel: 7 })
        .toBuffer(),
      fileName
    );
    formData.append("cacheControl", 3600);

    res.json(
      await fetch(
        `${process.env.NEXT_PUBLIC_SUPABASE_URL}/storage/v1/object/images/${fileName}`,
        {
          method: "POST",
          headers: {
            apikey,
            Authorization: `Bearer ${apikey}`,
          },
          body: formData as any,
        }
      ).then((response) => response.json())
    );
  } catch (err) {
    res.status(500).json({ error: err.message });
  }
}

```

**Reply by coderinblack08** [reply]
thanks @easy-street ! also, is it possible to store supabase objects as URLs in a database? I think the expiry dates conflict with this. The reason I want no expiry is that getting a signed URL takes some time.

**Reply by easy-street** [reply]
i think signed URLs are the only option here currently. maybe a long expiry would work for your use case (as suggested in this comment: https://github.com/supabase/supabase/discussions/1462#discussioncomment-707401)?

**Reply by coderinblack08** [reply]
@easy-street AWS S3 seems to put a 1 week limit on the expiry date for signed URLs so I assume this applies to supabase too since they use S3. Do you know if this is this true?

**Reply by easy-street** [reply]
i don't believe so. the referenced [comment](https://github.com/supabase/supabase/discussions/1462#discussioncomment-707401) certainly indicates there is no such limitation:

>If you generate a signed URL with an expiry of 1 month, it will continue to be valid for 1 month from the point you generated the signed URL. You can set a long expiry and put this in your website or mobile app.

**Reply by coderinblack08** [reply]
Lol imma make my expiration date like in 1000 years
